File: RaftSwift/Sources/RaftCore/Utils/ReplicationTracker.swift
/// Tracks the replication of log entries to peers.
actor ReplicationTracker {
    /// The number of peers that need to have replicated the log entries.
    let majority: Int

    /// The set of peers that have successfully replicated the log entries.
    var successful: Set<Int> = []

    /// The set of continuations to be resumed when the majority of peers have replicated the log entries.
    var continuations: [CheckedContinuation<Void, Never>] = []

    /// Initializes a new replication tracker.
    ///
    /// - Parameter majority: The number of peers that need to have replicated the log entries.
    init(majority: Int) {
        self.majority = majority
    }

    /// Marks a peer as successful.
    ///
    /// - Parameter id: The ID of the peer that has successfully replicated the log entries.
    func markSuccess(id: Int) {
        guard !successful.contains(id) else { return }
        successful.insert(id)
        if successful.count >= majority {
            for cont in continuations {
                cont.resume()
            }
            continuations.removeAll()
        }
    }

    /// Waits for the majority of peers to have replicated the log entries.
    func waitForMajority() async {
        if successful.count >= majority {
            return
        }

        await withCheckedContinuation { cont in
            if successful.count >= majority {
                cont.resume()
            } else {
                continuations.append(cont)
            }
        }
    }

    /// Returns whether a peer has successfully replicated the log entries.
    ///
    /// - Parameter id: The ID of the peer to check.
    func isSuccessful(id: Int) -> Bool {
        successful.contains(id)
    }

    /// Returns the set of peers that have successfully replicated the log entries.
    func getSuccessfulPeers() -> Set<Int> {
        successful
    }
}

File: RaftKotlin/raft-node/src/main/kotlin/raft/core/utils/ReplicationTracker.kt
package raft.core.utils

import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/** Tracks the replication of log entries to peers. */
class ReplicationTracker(
        /** The number of peers that need to have replicated the log entries. */
        private val majority: Int
) {
    /** The mutex for thread-safe access to the state. */
    private val mutex = Mutex()
    /** The set of peers that have successfully replicated the log entries. */
    private val successful: MutableSet<Int> = mutableSetOf()
    /**
     * The set of continuations to be resumed when the majority of peers have replicated the log
     * entries.
     */
    private val continuations: MutableList<CompletableDeferred<Unit>> = mutableListOf()

    /**
     * Marks a peer as successful.
     *
     * @param id The id of the peer that has successfully replicated the log entries.
     */
    suspend fun markSuccess(id: Int) {
        mutex.withLock {
            if (!successful.add(id)) {
                return
            }
            if (successful.size >= majority) {
                for (continuation in continuations) {
                    continuation.complete(Unit)
                }
                continuations.clear()
            }
        }
    }

    /** Waits for the majority of peers to have replicated the log entries. */
    suspend fun waitForMajority() {
        mutex.withLock {
            if (successful.size >= majority) {
                return
            }

            val continuation = CompletableDeferred<Unit>()
            continuations.add(continuation)
            mutex.unlock() // Avoid deadlock
            continuation.await()
            mutex.lock()
        }
    }

    /**
     * Returns whether a peer has successfully replicated the log entries.
     *
     * @param id The id of the peer to check.
     * @return True if the peer has successfully replicated the log entries, false otherwise.
     */
    suspend fun isSuccessful(id: Int): Boolean {
        return mutex.withLock { id in successful }
    }

    /**
     * Returns the set of peers that have successfully replicated the log entries.
     *
     * @return The set of peers that have successfully replicated the log entries.
     */
    suspend fun getSuccessfulPeers(): Set<Int> {
        return mutex.withLock { successful.toSet() }
    }
}
